### ISSUE TITLE
How to show hints for tabs?

### ISSUE THREAD

**Issue by nuanjanP** [tridactyl/tridactyl#4854]
Hi. I was wondering if tridactyl has a hint mode but for the tabs themselves, like [this feature in Surfingkeys](https://github.com/brookhong/Surfingkeys#switch-tabs). This would help me a lot visually when switching tabs. I tried looking at the tutorial and the github issues but I guess I missed something.

Thank you in advance :)

**Comment by bovine3dom** [comment]
No, we don't have that. I've never heard of that feature before :)

We have the `b` "omnibar" style switching but not hints.

I don't think we'll add it any time soon as it would be a fair amount of work. In the meantime, you can use `[n]gt` to go to the n-th tab, e.g. `5gt` to go to the fifth tab.

**Comment by Chic-Tweetz** [comment]
The tab / taball commands are great, but this intrigued me and I did manage to get a command working thanks to some of the handy stuff Tridactyl offers (tri.browserBg.tabs in particular!)


https://github.com/tridactyl/tridactyl/assets/57910448/93434bff-9fee-497f-86b2-96152f9d92e7


It's a bit of a mess and I've given up wrestling with CSS for now, but I can share it when it's neatened up a bit if you like :)

**Comment by Chic-Tweetz** [comment]
Hey, I've tinkered with this some more if you're interested. 

[ready to copy & paste into Tridactyl's command line](https://github.com/Chic-Tweetz/tridactyl-stuff/blob/main/hinttabs)

[script with line breaks so you can see what's going on](https://github.com/Chic-Tweetz/tridactyl-stuff/blob/main/hinttabs-readable)

**Comment by nuanjanP** [comment]
@bovine3dom  @Chic-Tweetz  Thank you very much! Your suggestions work great 😃

By the way, as for the code kindly given by @Chic-Tweetz :

> Hey, I've tinkered with this some more if you're interested.
> 
> [ready to copy & paste into Tridactyl's command line](https://github.com/Chic-Tweetz/tridactyl-stuff/blob/main/hinttabs)
> 
> [script with line breaks so you can see what's going on](https://github.com/Chic-Tweetz/tridactyl-stuff/blob/main/hinttabs-readable)

In the second link, line 8, I'm not sure but for me I had to change the second word `hinttaballts` needs to be `hinttaball` in order for line 13 (`bind T hinttaball`) to work. Not a big problem though. 😄 

Meanwhile, I stumbled upon this application (for Windows) called [Fluent Search](https://fluentsearch.net) which has a "[screen search](https://fluentsearch.net/posts/fluent-search-features-guide)" mode that is in essence tridactyl's hint mode but for the entire screen. While tridactyl is still better inside the browsing area, this application can hint other stuff currently not covered by tridactyl, like the tabs, items in Firefox's bookmark toolbar. Not sure it tridactyl can already do this, but Fluent Search also supports right clicking/double clicking/moving mouse cursor to hinted position, which can help with elements on the page that needs hovering (not clicking) to activate.

But anyway, having a customizable tab switcher inside tridactyl is still very welcomed. Thank you again for your time!

**Comment by Chic-Tweetz** [comment]
@nuanjanP 
> 
> In the first link, line 8, I'm not sure but for me I had to change the second word `hinttaballts` needs to be `hinttaball` in order for line 13 (`bind T hinttaball`) to work. Not a big problem though. 😄
> 
Whoops! I was fixing some stuff and had a couple of commands going to compare them. Must have neglected to correct the name for the paste. 

I'll have a look at Fluent, sounds like a good tool for a Tridactyl fan :)

As it happens, I have an AHK script (and a Hammerspoon one for my macbook) to move the mouse (or right click and stuff) that can be used Tridactyl's execute command. So not directly possible in Tridactyl, but doable!

**Comment by bovine3dom** [comment]
Thanks for sharing your code, @Chic-Tweetz. I would like to integrate it into Tridactyl but I think it leaks the names of the tabs to the web page.

I think we'd need to put it in an iframe to protect it, like we do with the command line. I don't think the ShadowDOM would be enough. But that means that our hinting code would need to be rewritten to support iframes - see #315.

Alternatively, we could integrate your snippet into Tridactyl but leave it unbound with a big health warning on the command's help saying that it could give the details of all your tabs to the page you activate it on.

**Comment by Chic-Tweetz** [comment]
@bovine3dom Ah, yes I don't usually worry about that kind of thing when I'm just making scripts for myself.
I'll have a look at iframes. Might it be possible to use messaging with an iframe and hintable elements that sit outside of it?
I'm happy for it to be integrated with a warning, plus if anyone takes it and improves on it that's all fine by me.

**Comment by Chic-Tweetz** [comment]
> 
> I think we'd need to put it in an iframe to protect it, like we do with the command line. I don't think the ShadowDOM would be enough. But that means that our hinting code would need to be rewritten to support iframes - see #315.
>

I don't know if this could be a solution for this but I did try out iframes with messaging. This way might keep tab names within the iframe while letting hintable elements get tab IDs.
It's not a working solution, just some commands to paste with an iframe html file.

**Comment by Skillmon** [comment]
Just a small addition to

> you can use `[n]gt` to go to the n-th tab

together with `tabcounters` you practically got hints for your tabs "always on", as all the tabs will display the number for `gt` in front of the name.

**Comment by bovine3dom** [comment]
@Chic-Tweetz thanks, that's a cool trick. I'm a bit reluctant to integrate it as it feels like it's adding a lot of complexity when we need to fix #315 anyway so that hinting (and binds) can be used in all frames

**Comment by Chic-Tweetz** [comment]
@bovine3dom oh yeah it's a lot more complicated for sure. I have made an iframe version that works as one js command, so at least for this tab hinting idea it can be done with that extra layer of complexity working out where the iframe's elements will end up.

**Comment by glacambre** [comment]
Perhaps the solution is to have one frame per tab element, the page would be unable to read the frame's content but Tridactyl would be able to know what tab to select based on what frame is selected. The remaining tricky part would be to let iframes know what content they should display to the user, but perhaps having them ask the content script for this information on creation and having the content script create them in sequence could be the solution?

**Comment by glacambre** [comment]
>creates an html document as a string then makes a blob with it, which Firefox is happy to let be the source of an iframe

Wouldn't the page be able to read the blob, and thus the iframe's content, through the iframe's src? I'm not sure the need for iframe<->content script can be avoided unfortunately :/.

**Comment by Chic-Tweetz** [comment]
> 
> Wouldn't the page be able to read the blob, and thus the iframe's content, through the iframe's src? I'm not sure the need for iframe<->content script can be avoided unfortunately :/.

🤦 Yes you're right. I'll have to stop trying to outsmart iframes.

**Comment by Chic-Tweetz** [comment]
> Hey, I've tinkered with this some more if you're interested.
> 
> [ready to copy & paste into Tridactyl's command line](https://github.com/Chic-Tweetz/tridactyl-stuff/blob/main/hintable-tabs/commands/hinttabs)
> 
> [script with line breaks so you can see what's going on](https://github.com/Chic-Tweetz/tridactyl-stuff/blob/main/hintable-tabs/js/hinttabs.js)

Hello again, I've moved things around in this repo and updated the links here accordingly. The script itself is different to before. It puts everything in a shadow dom so has a bunch of css rules in it & I put new tab, close tab and search (taball) buttons in.

This is a bit silly but I also made [this script](https://github.com/Chic-Tweetz/TridactylHintTabs/blob/main/tridactyl/tridactyl_script.js) (& [minified commands](https://github.com/Chic-Tweetz/TridactylHintTabs/tree/main/tridactyl/commands)) which puts the tab info into an iframe with a github.io page as its source. (It also works if set to use a localhost source but I wanted to make one that would work with just a copy/paste into the command line).

**Comment by trapd00r** [comment]
Great work @Chic-Tweetz , use it and love it!

**Comment by Chic-Tweetz** [comment]
I recently found that creating a blob URL through Tridactyl's command line iframe's contentWindow gives a "blob:moz-extension://..." URL, eg running:
```
:js
const blob = new Blob();
const blurl = document.getElementById("cmdline_iframe").contentWindow.URL.createObjectURL(blob);
const iframe = document.createElement("iframe");
iframe.id = "obscuredIframe";
iframe.src = blurl;
document.body.appendChild(iframe);
```
then attempting to access the contentWindow or contentDocument shows they are restricted / null.
```
document.getElementById("obscuredIframe").contentDocument // null
document.getElementById("obscuredIframe").contentWindow   // restricted
```

[So I've made another script](https://github.com/Chic-Tweetz/tridactyl-stuff/blob/main/hinttabs-obscured/hinttabs.js) which takes advantage of that, plus it should take colours from Tridactyl themes, + added outline colours according to containers, and tab prefixes like :tab / :taball.

Instead of making a script each for tab/taball, the linked script is to create an alias which checks for a "-a" flag, so :hinttabs -a displays tabs from every window and :hinttabs just the current window.

**Comment by bovine3dom** [comment]
That looks cool @Chic-Tweetz , I'll look at it next week if I find the time.

I think your technique should mean that people can't just reconstruct data from the iframe source URL, right?

**Comment by Chic-Tweetz** [comment]
@bovine3dom yes, or at least I haven't found a way to. The iframe can be accessed through Tridactyl, but not directly.
